zebra-chain/src/block/arbitrary.rs

```diff
@@ -326,14 +328,79 @@ impl Block {
+                    let mut transaction = (*transaction).clone();
+                    let mut new_inputs = Vec::new();
+
+                    for mut input in transaction.inputs_mut().drain(..) {
+                        if let PrevOut {
+                            ref mut outpoint, ..
+                        } = input
+                        {
+                            // take a UTXO if available
+                            if utxos.remove(outpoint) {
+                                new_inputs.push(input);
+                            } else if let Some(arbitrary_utxo) = utxos.clone().iter().next() {
+                                *outpoint = *arbitrary_utxo;
+                                utxos.remove(arbitrary_utxo);
+                                new_inputs.push(input);
+                            }
+                            // otherwise, drop the invalid input, it has no UTXOs to spend
+                        } else {
+                            // preserve coinbase inputs
+                            new_inputs.push(input);
+                        }
+                    }
+
+                    // delete invalid inputs
+                    *transaction.inputs_mut() = new_inputs;
+
+                    // keep transactions with valid input counts
+                    // coinbase transactions will never fail this check
+                    // this is the input check from `has_inputs_and_outputs`
+                    if !transaction.inputs().is_empty()
+                        || transaction.joinsplit_count() > 0
+                        || transaction.sapling_spends_per_anchor().count() > 0
+                        || (transaction.orchard_actions().count() > 0
+                            && transaction
+                                .orchard_flags()
+                                .unwrap_or_else(orchard::Flags::empty)
+                                .contains(orchard::Flags::ENABLE_SPENDS))
+                    {
+                        // add the created UTXOs
+                        // these outputs can be spent from the next transaction in this block onwards
+                        // see `new_outputs` for details
+                        let hash = transaction.hash();
+                        for output_index_in_transaction in 0..transaction.outputs().len() {
+                            utxos.insert(transparent::OutPoint {
+                                hash,
+                                index: output_index_in_transaction.try_into().unwrap(),
+                            });
+                        }
+
+                        // and keep the transaction
+                        new_transactions.push(Arc::new(transaction));
+                    }
+                }
+
+                // delete invalid transactions
+                block.transactions = new_transactions;
+
+                // TODO: fixup the history and authorizing data commitments, if needed
```

Comment:
@conradoplg @dconnolly we might need similar fixup code for the anchors ⬆️ 
Or test-only methods like `inputs_mut`, which let us edit test blocks or the state ⬇️